### ISSUE TITLE
constructor initializer list — make expressions able to use earlier names in the list

### ISSUE THREAD

**Issue by SteveAlexander** [dart-lang/language#1394]
In a constructor's initializer list, I'd like expressions to be able to use variables from the same list that were initialized earlier in the list.

I've read https://github.com/dart-lang/sdk/issues/26655
and https://github.com/dart-lang/sdk/issues/15346

So, I'm glad that it's possible to do this:

```
class C {
  final int x;
  final int y;
  C(this.x) : y = x + 1;
}
```

I would also like to be able to do this, which presently gives me an error "Only static members can be accessed in initializers."

```
class C {
  final int x;
  final int y;
  final int z;
  C(this.x)
    : y = x + 1,
      z = y + 1;
}
```

I see in https://github.com/dart-lang/sdk/issues/28950 that something similar was requested, and rejected with the rationale "We will definitely not make the initializer list have access to the real object.".

However, I don't want that as such. I'd like to have expressions later down the initializer list able to use initialized earlier in the list.

I meet this from time to time during Flutter development where I want to create a `StatelessWidget`, and so keep all of its members `final`, and doing this as part of the initializer for something related to initializing the widget seems to be cleaner than creating a method or getter to do the same.

Also, it feels to me intuitive that it should work this way.

* Dart SDK Version (`dart --version`)
Dart VM version: 2.1.0-dev.4.0.flutter-cd9a42239f (Fri Sep 7 21:08:23 2018 +0000) on "macos_x64"

**Comment by larssn** [comment]
@rich-j Have you considered `late final` now that we have null safety?

Dartpad example:
```
class LateTest {
  LateTest() : stream1 = Stream.periodic(Duration(seconds: 1), (val) => val).map((val) => val).asBroadcastStream() {    
    stream2 = stream1.map((val) => val.toString() + ' cookie' + (val > 1 ? 's' : ''));
    stream3 = stream1.map((val) => val.toString() + ' l milk');
  }
  
  final Stream<int> stream1;
  late final Stream<String> stream2;
  late final Stream<String> stream3;
  
  void start() {
    stream2.listen(print);
    stream3.listen(print);
  }
}

main() {
  LateTest()..start();
}
```

I realize this introduces a runtime check for each `late final` variable, and I'm not sure if it has any performance implications. Maybe someone can elaborate on that.

**Comment by rich-j** [comment]
@larssn Yes, we are happily mostly using `late final` now instead of constructor initializers. `late final` allows the definition to be provided with the declaration. To expand on your example...

```dart
class LateTest {
  
  final int count;
  final Duration waitTime;
  
  LateTest({required this.count, this.waitTime = const Duration(seconds: 1)});
  
  late final Stream<int> stream1 = Stream.periodic(waitTime, (val) => val).take(count).asBroadcastStream();
  late final Stream<String> stream2 = stream1.map((val) => "$val cookie${val > 1 ? 's' : ''}");
  late final Stream<String> stream3 = stream1.map((val) => "$val l milk");
  
  void start() {
    stream2.listen(print);
    stream3.listen(print);
  }
}

main() {
  LateTest(count: 3)..start();
}
```

We mostly code using immutable objects that depend on provided external values (e.g. from network or database) that other values are derived from. 

As for performance it's always a tradeoff. Yes, `late final` will have to check every access but in most code paths that should be a negligible (i.e. don't prematurely optimize). In the above example using declaration initializers, if you only access `stream2` the initialization of `stream3` wouldn't need to happen.

**Comment by sebastianhaberey** [comment]
+1

From a new Dart user's perspective, this definitely seems like a kink in the language. 

According to the [documentation](https://dart.dev/guides/language/language-tour#late-variables), the _late_ keyword is intended for variables that are not initialized at the time of their declaration and variables that need lazy initialization. Both of these criteria don't fit well; what we'd really like to have is initialization at the time of declaration for variables that depend on each other.

For me it comes down to a question of wording: "late" just doesn't fit for what I'm trying to achieve. It just feels wrong.

**Comment by AlexanderFarkas** [comment]
@sebastianhaberey 
I use it this way:
```dart
late final field = this.instanceField + 2;
```

**Comment by lrhn** [comment]
Worth noticing that the suggestions to allow fields to access other fields,
```dart
class C {
  final int x = 1;
  final int y = x + 1;
}
```
are *less* viable with the introduction of the augmentation feature currently being design.

The augmentation feature allows an augmenting declaration to override/augment the getter and setter of a field declaration, which means that `x` in `final int y = x + 1;` doesn't refer to a *variable*, but to a getter that can do anything, and which cannot be allowed to run before the object is completely initialized.

That was actually already the case before. If `C` was followed by:
```dart
class D {
  int otherX = 42;
  int get x => someRandom.nextBool() ? super.x : otherX;
}
```
it's not *clear* that `final int y = x + 1;` should refer to the value of the `x` field initializer, and not to the instance getter. It would have to be defined that way, and then
```dart
 final int y = x + 1;
 late final int z = x + 1;
```
would behave differently for the exact same code. That's not great.

With augmentations, it can be done inside the same class:
```dart
class C {
  final someRandom = Random();
  int otherX = 42;
  final int x = 1;
  final int y = x + 1;
  augment get x => someRandom.nextBool ? augmented : otherX;
}
```
The reference to `x` in `final int y = x + 1;` *should* refer be a `this.x` member access, which means it hits the getter and cannot be allowed in an initializer.

(I also want to refer to prior values in initializer lists. As late as today I did a workaround with a second constructor, just to be able to initialize what this feature would allow you to write as just
```dart
 : currentTime = (someExpression).millisecondsSinceEpoch, nextTime = currentTime - 1;
```
So there is demand!)
